packet/eip_controlplane_reconciliation.go

```diff
@@ -0,0 +1,131 @@
+}
+
+func (m *controlPlaneEndpointManager) Reconciler() func(nodes []*v1.Node, remove bool) error {
+	return func(nodes []*v1.Node, remove bool) error {
+		klog.V(2).Info("controlPlaneEndpoint.reconcile: new reconciliation")
+		ipList, _, err := m.ipResSvr.List(m.projectID)
+		if err != nil {
+			return err
+		}
+		controlPlaneEndpoint := ipReservationByTags([]string{m.eipTag}, ipList)
+		if controlPlaneEndpoint == nil {
+			// IP NOT FOUND nothing to do here.
+			klog.Error("Elastic IP not found. Please verify you have one with the expected tag: %s", m.eipTag)
+			return err
+		}
+		klog.Infof("healthcheck elastic ip %s", controlPlaneEndpoint.Address)
+		req, err := http.NewRequest("GET", "https://"+controlPlaneEndpoint.Address+":6443/healthz", nil)
+		if err != nil {
+			klog.Errorf("healthcheck failed for control plane endpoint %s. err \"%s\". Checking the next one...", controlPlaneEndpoint.Address, err)
+			return err
+		}
+		resp, err := m.httpClient.Do(req)
+		if err != nil || resp.StatusCode != http.StatusOK {
+			if err := m.reassign(context.Background(), nodes, controlPlaneEndpoint); err != nil {
+				klog.Errorf("error reassigning control plane endpoint to a different device. err \"%s\"", err)
+				return err
+			}
+		}
+		defer resp.Body.Close()
+		return nil
+	}
+}
+
+func (m *controlPlaneEndpointManager) reassign(ctx context.Context, nodes []*v1.Node, ip *packngo.IPAddressReservation) error {
+	klog.V(2).Info("controlPlaneEndpoint.reassign")
+	for _, node := range nodes {
+		addresses, err := m.instances.NodeAddresses(ctx, types.NodeName(node.Name))
+		if err != nil {
+			return err
+		}
+		// It does not matter if the IP is external or internal. We can reach both.
+		a := addresses[0]
+
+		klog.Infof("healthcheck node %s", a.Address)
+		req, err := http.NewRequest("GET", "https://"+a.Address+":6443/healthz", nil)
+		if err != nil {
+			klog.Errorf("healthcheck failed for node %s. err \"%s\"", node.Name, err)
+			continue
+		}
+		resp, err := m.httpClient.Do(req)
+
+		if err != nil {
+			continue
+		}
+
+		// We have a healthy node, this is the candidate to receive the EIP
+		if resp.StatusCode == http.StatusOK {
+			deviceID, err := m.instances.InstanceID(ctx, types.NodeName(node.Name))
+			if err != nil {
+				return err
+			}
+			if _, err := m.deviceIPSrv.Unassign(ip.ID); err != nil {
+				return err
+			}
+			if _, _, err := m.deviceIPSrv.Assign(deviceID, &packngo.AddressStruct{
+				Address: ip.Address,
+			}); err != nil {
+				return err
+			}
+			klog.V(2).Infof("control plane endpoint assigned to new device %s", node.Name)
+		}
+
+	}
+	return errors.New("ccm didn't find a good candidate for IP allocation. Cluster is unhealthy")
+}
+
+func newControlPlaneEndpointManager(eipTag, projectID string, deviceIPSrv packngo.DeviceIPService, ipResSvr packngo.ProjectIPService, i cloudInstances) *controlPlaneEndpointManager {
+	return &controlPlaneEndpointManager{
+		httpClient: &http.Client{
+			Timeout: time.Second * 5,
```

Comment:
I think we will want control over this timeout, perhaps as a command-line parameter.  Can we define this in a const for now?

Reply:
We definitely want control. If you can get it in quickly @gianarb then great, but it is important we get this done before the weekend, so if it holds us up even a few mins, we can do a separate PR on it after.